File: app/Models/TemplateCategory.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Factories\HasFactory;

class TemplateCategory extends Model
{
    use SoftDeletes, HasFactory;

    protected $fillable = [
        'name',
        'description',
        'parent_category_id',
        'order',
    ];

    protected $casts = [
        'created_at' => 'datetime',
        'updated_at' => 'datetime',
        'deleted_at' => 'datetime',
    ];

    /**
     * Get the parent category if this is a subcategory
     */
    public function parentCategory(): BelongsTo
    {
        return $this->belongsTo(TemplateCategory::class, 'parent_category_id');
    }

    /**
     * Get the child categories of this category
     */
    public function childCategories(): HasMany
    {
        return $this->hasMany(TemplateCategory::class, 'parent_category_id')
            ->orderBy('order');
    }

    /**
     * Get all templates in this category
     */
    public function templates(): HasMany
    {
        return $this->hasMany(Template::class, 'category_id')
            ->orderBy('order');
    }
}

File: src/main/kotlin/models/TemplateCategory.kt
package models

import com.expediagroup.graphql.server.extensions.getValueFromDataLoader
import graphql.schema.DataFetchingEnvironment
import kotlinx.serialization.Serializable
import kotlinx.datetime.LocalDateTime
import dataloaders.TemplateCategoryDataLoader
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import services.TemplateCategoryService
import tables.CategorySpecialType
import util.now
import java.util.concurrent.CompletableFuture
import kotlin.getValue


object TemplateCategoryHelper : KoinComponent {
    private val service: TemplateCategoryService by inject()
}

@Serializable
data class TemplateCategory(
    val id: Int = 0,
    val name: String,
    val description: String? = null,
    val parentCategoryId: Int? = null,
    val order: Int? = null,
    val categorySpecialType: CategorySpecialType? = null,
    val deletedAt: LocalDateTime? = null,
    val createdAt: LocalDateTime = now(),
    val updatedAt: LocalDateTime = now()
) {
    fun parentCategory(dataFetchingEnvironment: DataFetchingEnvironment): CompletableFuture<TemplateCategory?> {
        return if (parentCategoryId != null) {
            dataFetchingEnvironment.getValueFromDataLoader(TemplateCategoryDataLoader.dataLoaderName, parentCategoryId)
        } else CompletableFuture.completedFuture(null)
    }
}

data class CreateTemplateCategoryInput(
    val name: String,
    val description: String?,
    val parentCategoryId: Int?,
)

data class UpdateTemplateCategoryInput(
    val id: Int,
    val name: String,
    val description: String?,
    val parentCategoryId: Int?,
    val order: Int
)
